pallene/checker.lua

```diff
@@ -465,6 +465,35 @@ function FunChecker:check_stat(stat)
             stat.decl._name = self.p.symbol_table:find_symbol(stat.decl.name)
             self:check_stat(stat.block)
         end)
+        
+    elseif tag == "ast.Stat.ForIn" then
+        local func = self:check_exp_synthesize(stat.exp)
+
+        if func._tag ~= "ast.Exp.CallFunc" then
+            type_error(func.loc,
+                "expected ipairs function but found %s in for-loop",
+                types.tostring(func._type),
+                func.name)
+        end
+
+        if func.exp._tag ~= "ast.Exp.Var" then -- check if arg is an array
+            type_error(func.exp.loc,
+                "expected table or array in ipairs",
+                func.exp.name)
+        end
```

Comment:
done

Reply:
It does not suffice to check if the variable is named `ipairs`, because the programmer can create other variables named ipairs. We have to check the `_name` to confirm that it is actually the builtin ipairs.

    local ipairs: integer = 17
    for i, x in ipairs(xs) do

Furthermore, I think we might not need to check that `arr` has type array here. The `synthesize(stat.exp)` should have already done that.

Reply:
I, too, thought that ```synthesize(stat.exp)``` checks if arguments if array, it did actually. But I did smth wrong and now it does not check it properly

Reply:
> But I did smth wrong and now it does not check it properly

What exactly isn't working?

> type error: attempting to call a integer value

I think that error is right. In Lua we would also get a similar type error if we overwrite the value of `ipairs` with something else.

Reply:
Now, I think, I have fixed that